specifyweb/frontend/js_src/lib/components/TreeView/AddRank.tsx

```diff
@@ -0,0 +1,103 @@
+import React from 'react';
+
+import { commonText } from '../../localization/common';
+import { interactionsText } from '../../localization/interactions';
+import { treeText } from '../../localization/tree';
+import type { RA } from '../../utils/types';
+import { Button } from '../Atoms/Button';
+import { Label, Select } from '../Atoms/Form';
+import type {
+  AnyTree,
+  FilterTablesByEndsWith,
+  SerializedResource,
+} from '../DataModel/helperTypes';
+import { tables } from '../DataModel/tables';
+import { ResourceView } from '../Forms/ResourceView';
+import { Dialog } from '../Molecules/Dialog';
+
+export function AddRank<SCHEMA extends AnyTree>({
+  tableName,
+  treeDefinitionItems,
+}: {
+  readonly tableName: SCHEMA['tableName'];
+  readonly treeDefinitionItems: RA<
+    SerializedResource<FilterTablesByEndsWith<'TreeDefItem'>>
+  >;
+}): JSX.Element {
+  const [state, setState] = React.useState<'add' | 'initial' | 'parent'>(
+    'initial'
+  );
+
+  const [parentRank, setParentRank] = React.useState(
+    treeDefinitionItems[0].resource_uri
+  );
+
+  const treeDef = treeDefinitionItems[0].treeDef;
+
+  const treeResource = React.useMemo(() => {
+    const resource = new tables[treeDefinitionItems[0]._tableName].Resource();
+    resource.set('treeDef', treeDef);
+    return resource;
+  }, [tableName, treeDef]);
+
+  return (
+    <>
+      <Button.Icon
+        icon="plus"
+        title={treeText.addNewRank()}
+        onClick={() => setState('parent')}
+      />
+      {state === 'parent' && (
+        <Dialog
+          buttons={
+            <>
+              <Button.DialogClose>{commonText.cancel()}</Button.DialogClose>
+              <Button.Save
+                onClick={() => {
+                  treeResource.set('parent', parentRank);
+                  setState('add');
+                }}
+              >
+                {interactionsText.continue()}
+              </Button.Save>
+            </>
+          }
+          header={treeText.chooseParentRank()}
+          onClose={() => setState('initial')}
+        >
+          <Label.Block className="gap-2">
```

Comment:
(optional) use `<Form>` and `<Submit.Save>` here - that way, the Enter key would work automatically to submit the form

Reply:
mhhh kook good to know